### PR TITLE
[Feature] Add AAD Custom Auth Support & Setup Custom Auth Framework

### PR DESCRIPTION
**Pull request by Timothyw0**
New feature: Adding AAD custom auth support using the previously created code for Google and GitHub. The previous code has been refactored and generalized for future auth providers to be added as well. Google and Github auth has also been regression tested successfully.

Tested in private environment with custom AAD application using the following staticwebapp.config.json:
image

Successful /.auth/me call with AAD credentials:
image

Successful /.auth/me call with Github credentials:
image

Successful /.auth/me call with Google credentials:
image

### CODE REVIEW

File: src/core/constants.ts [hunk 1]
```diff
@@ -49,6 +49,44 @@ export const SWA_AUTH_CONTEXT_COOKIE = `StaticWebAppsAuthContextCookie`;
 export const SWA_AUTH_COOKIE = `StaticWebAppsAuthCookie`;
 export const ALLOWED_HTTP_METHODS_FOR_STATIC_CONTENT = ["GET", "HEAD", "OPTIONS"];
 
+// Custom Auth constants
+export const SUPPORTED_CUSTOM_AUTH_PROVIDERS = ["google", "github", "aad"];
+// Full name is required in staticwebapp.config.json's schema so we will normalize it to aad
+export const AAD_FULL_NAME = "azureActiveDirectory";
```
Comment (adrianhall) [inline src/core/constants.ts:55]: Change to entraId or similar - AAD is the old name - need new branding.

File: src/msha/auth/index.ts [hunk 2]
```diff
@@ -9,20 +9,20 @@ function getAuthPaths(isCustomAuth: boolean): Path[] {
   if (isCustomAuth) {
     paths.push({
       method: "GET",
-      // only match for providers with custom auth support implemented (github, google)
-      route: /^\/\.auth\/login\/(?<provider>github|google|dummy)\/callback(\?.*)?$/i,
+      // only match for providers with custom auth support implemented (github, google, aad)
+      route: /^\/\.auth\/login\/(?<provider>github|google|aad|dummy)\/callback(\?.*)?$/i,
```
Comment (adrianhall) [inline src/msha/auth/index.ts:13]: Can we construct this based on the extensible list of supported providers?

File: src/msha/auth/index.ts [hunk 3]
```diff
@@ -9,20 +9,20 @@ function getAuthPaths(isCustomAuth: boolean): Path[] {
...
-      route: /^\/\.auth\/login\/(?<provider>github|google|dummy)(\?.*)?$/i,
+      // only match for providers with custom auth support implemented (github, google, aad)
+      route: /^\/\.auth\/login\/(?<provider>github|google|aad|dummy)(\?.*)?$/i,
       function: "auth-login-provider-custom",
     });
     paths.push({
       method: "GET",
       // For providers with custom auth support not implemented, revert to old behavior
-      route: /^\/\.auth\/login\/(?<provider>aad|twitter|facebook|[a-z]+)(\?.*)?$/i,
+      route: /^\/\.auth\/login\/(?<provider>twitter|facebook|[a-z]+)(\?.*)?$/i,
```
Comment (adrianhall) [inline src/msha/auth/index.ts:25]: Does SWA support twitter/facebook any more?  If not, they are custom too.  Implement what is in SWA today.

Comment (Timothyw0) [reply src/msha/auth/index.ts:25]: Yes we do, that will be included in the next PR :)

File: src/core/constants.ts (hunk 1 above)
Comment (Timothyw0) [reply src/core/constants.ts:55]: I can only change the variable name to `ENTRAID_FULL_NAME` for now. The configuration json is still expecting `azureActiveDirectory` so until that is changed, I cannot re-brand the expected string.

**Review by mkarmark — APPROVED** [review]

**Pull request merged**